bcrembed/utils.py

```diff
@@ -32,23 +33,79 @@ def insert_space_every_other_except_cls(input_string):
-
+        inpath (str): The file path to the input data.
+        sequence_input (str): The type of sequence input, which can be one of ["H", "L", "HL"].
+        
     Returns:
-    DataFrame: The pivoted DataFrame.
+        pandas.DataFrame: Processed AIRR-seq data based on the sequence_input.
+        
+    Raises:
+        ValueError: If sequence_input is not one of ["H", "L", "HL"].
     """
+    allowed_sequence_input = ["H", "L", "HL"]
+    if sequence_input not in allowed_sequence_input:
+        raise ValueError("Input x must be one of {}".format(allowed_sequence_input))
+        
     data = pd.read_table(inpath)
-    colnames = ['cell_id', 'locus', 'consensus_count', 'sequence_vdj_aa']
-    data = data.loc[:, colnames]
     data.loc[:,'chain'] = data.loc[:,'locus'].apply(lambda x: 'H' if x == 'IGH' else 'L')
-    data = data.loc[data.groupby(['cell_id', 'chain'])['consensus_count'].idxmax()]
-    data = data.pivot(index='cell_id', columns='chain', values='sequence_vdj_aa')
-    data.loc[:,'HL'] = data.H + '<cls><cls>' + data.L
+    
+    if not 'cell_id' in data.columns:
+        data_type = 'bulk-only'
+    elif data['cell_id'].notna().all():
+        data_type = 'single-cell-only'
+    else:
+        data_type = 'mixed'
+
+    if data_type == 'bulk-only':
+        logging.info("No cell_id column detected. Processsing as bulk data.")
+        if sequence_input == 'HL':
+            raise ValueError('sequence_input invalid for bulk mode.')
+        else: 
+            colnames = ['sequence_id', 'sequence_vdj_aa']
+            data = data.loc[data.chain == sequence_input, colnames]      
+            
+    elif data_type == 'single-cell-only':
+        logging.info("Processing single-cell BCR data...")
+        if sequence_input == "HL":
+            logging.info("Concatenating heavy and light chain per cell...")
+            data = concatenate_HL(data)
+        else:
+            colnames = ['cell_id', 'sequence_vdj_aa']
+            data = data.loc[data.chain == sequence_input, colnames]
+        
+    else:
+        logging.info("Missing values in cell_id column. Processing as mixed bulk and single-cell BCR data...")
+        if sequence_input == "HL":
+            logging.info("Concatenating heavy and light chain per cell...")
+            data = data.loc[data.cell_id.notna(),]
+            data = concatenate_HL(data)
+        else:
+            colnames = ['sequence_id', 'cell_id', 'sequence_vdj_aa']
+            data = data.loc[data.chain == sequence_input, colnames]
+        
     return data
 
+def concatenate_HL(data):
+    """
+    Concatenates heavy and light chain per cell and returns a pandas DataFrame.
+    
+    Parameters:
+        data (pandas.DataFrame): Input data containing information about heavy and light chains.
+        
+    Returns:
+        pandas.DataFrame: Dataframe with concatenated heavy and light chains per cell.
+    """
+    colnames = ['cell_id', 'locus', 'consensus_count', 'sequence_vdj_aa']
```

Comment:
I would add a check before that that these needed columns exist, and if not throw an error. Also I would let the user specify a different column name for the `sequence_vdj_aa` column, as this is not a standard AIRR column field